### PR TITLE
Weather Warnings

### PR DESCRIPTION
**Pull request by Geevies**
* Atmospheric probes will now relay changes in weather when deployed. Analog radios can be tuned into the weather broadcast and receive it, provided they're in the same sector / connected z-levels.
* Fixed weather transitions to rainy / icy weather not working.
* Changed rainy / icy weather to have a blue color, reducing eye strain.

image
image-1
image

### CODE REVIEW

**Comment by Geevies** [comment]
!review

File: code/__DEFINES/dcs/signals/signals_weather.dm [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// Signals related to weather and weather transitions
+// Sent primarily from weather_fsm.dm
+
+#define COMSIG_GLOB_Z_WEATHER_CHANGE "!z_weather_change"
+#define COMSIG_GLOB_Z_WEATHER_BROADCAST "!z_weather_broadcast"
```
Comment (FluffyGhoster) [inline code/__DEFINES/dcs/signals/signals_weather.dm:5]: DMDoc, also why does this start with an exclamation mark? That makes it look like it's saying it's not what the text says

File: code/game/objects/items/lore_radio.dm [hunk 2]
```diff
@@ -24,41 +27,56 @@
...
-		var/list/hearers = get_hearers_in_view(7, src)
-		var/list/clients_in_hearers = list()
-		for(var/mob/mob in hearers)
-			if(mob.client)
-				clients_in_hearers += mob.client
-		if(length(clients_in_hearers))
-			INVOKE_ASYNC(src, TYPE_PROC_REF(/atom/movable, animate_chat), radio_message, null, FALSE, clients_in_hearers, 2 SECONDS)
+		output_spoken_message(radio_message)
+	else
+		audible_message("\The <b>[src]</b> only emits white noise...")
```
Comment (FluffyGhoster) [inline code/game/objects/items/lore_radio.dm:67]: SPAN_BOLD

File: code/game/objects/items/lore_radio.dm [hunk 3]
```diff
@@ -24,41 +27,56 @@
...
+
+/obj/item/lore_radio/proc/relay_weather_broadcast(var/datum/source, var/z_level, var/singleton/state_transition/weather/weather_transition, var/time_to_transition, var/broadcast_message)
```
Comment (FluffyGhoster) [inline code/game/objects/items/lore_radio.dm:69]: DMDoc

File: code/game/objects/objs.dm [hunk 4]
```diff
@@ -301,3 +301,14 @@
...
+/obj/proc/output_spoken_message(var/message, var/message_verb = "transmits", var/display_overhead = TRUE, var/overhead_time = 2 SECONDS)
+	audible_message("\The <b>[src.name]</b> [message_verb], \"[message]\"")
+	if(display_overhead)
+		var/list/hearers = get_hearers_in_view(7, src)
+		var/list/clients_in_hearers = list()
+		for(var/mob/mob in hearers)
+			if(mob.client)
+				clients_in_hearers += mob.client
+		if(length(clients_in_hearers))
+			INVOKE_ASYNC(src, TYPE_PROC_REF(/atom/movable, animate_chat), message, null, FALSE, clients_in_hearers, overhead_time)
```
Comment (FluffyGhoster) [inline code/game/objects/objs.dm:314]: Why are you not using `visible_message` here?

File: code/game/objects/structures/survey_probe.dm [hunk 5]
```diff
@@ -69,11 +75,47 @@
 	anchored = TRUE
 	density = TRUE
 	icon_state = "[initial(icon_state)]_deployed"
+	if(survey_type == SURVEY_TYPE_ATMOSPHERIC)
+		RegisterSignal(SSdcs, COMSIG_GLOB_Z_WEATHER_CHANGE, PROC_REF(relay_weather_change))
+		addtimer(CALLBACK(src, PROC_REF(read_initial_weather)), 2 SECONDS)
+		output_spoken_message("Initializing weather detection subsystem...")
+
+/obj/structure/survey_probe/proc/read_initial_weather()
```
Comment (FluffyGhoster) [inline code/game/objects/structures/survey_probe.dm:83]: DMDoc

File: code/game/objects/structures/survey_probe.dm [hunk 6]
```diff
@@ -69,11 +75,47 @@
...
+		output_spoken_message("Reading current weather conditions as \"[current_weather_state.name]\".")
 
 /obj/structure/survey_probe/proc/undeploy()
 	anchored = FALSE
 	density = FALSE
 	icon_state = initial(icon_state)
+	if(survey_type == SURVEY_TYPE_ATMOSPHERIC)
+		UnregisterSignal(SSdcs, COMSIG_GLOB_Z_WEATHER_CHANGE)
+
+/obj/structure/survey_probe/proc/relay_weather_change(var/datum/source, var/z_level, var/singleton/state_transition/weather/weather_transition, var/time_to_transition)
```
Comment (FluffyGhoster) [inline code/game/objects/structures/survey_probe.dm:102]: DMDoc

File: code/modules/weather/weather_fsm.dm [hunk 7]
```diff
@@ -8,3 +8,15 @@
...
 		return pick(transitions) //pickweight(transitions)
 	return ..()
+
+/datum/state_machine/weather/handle_next_transition(var/obj/abstract/weather_system/holder_instance, var/singleton/state_transition/chosen_transition)
+	holder_instance.transitioning_weather = TRUE
+	var/time_to_transition = (rand(1, 3) MINUTES) + (rand(-30, 30) SECONDS)
+	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_Z_WEATHER_CHANGE, holder_instance.z, chosen_transition, time_to_transition)
+	addtimer(CALLBACK(src, PROC_REF(finalize_weather_transition), holder_instance, chosen_transition), time_to_transition)
+
+/datum/state_machine/weather/proc/finalize_weather_transition(var/obj/abstract/weather_system/holder_instance, var/singleton/state_transition/chosen_transition)
```
Comment (FluffyGhoster) [inline code/modules/weather/weather_fsm.dm:18]: DMDoc

File: code/game/objects/objs.dm (hunk 4 above)
Comment (Geevies) [reply code/game/objects/objs.dm:314]: it's an audible message, and this'll put the floating message "speech bubble" above it

File: code/__DEFINES/dcs/signals/signals_weather.dm (hunk 1 above)
Comment (Geevies) [reply code/__DEFINES/dcs/signals/signals_weather.dm]: image

**Comment by Geevies** [comment]
!review

**Review by Arrow768 — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**